H4HIP: Plugin-first architecture (#388)
helm/community: hips/hip-9999.md

---
hip: 9999
title: "H4HIP: Plugin-first architecture"
authors: [ "Scott Rigby <scott@r6by.com>", "George Jenkins <gvjenkins@gmail.com>" ]
created: "2025-02-17"
type: "feature"
status: "draft"
helm-version: "4"
---

## Glossary

WIP

> - **Legacy CLI plugins**: these are the traditional subprocess plugins that Helm 3 would call as user-specified binaries, and which weould extend the Helm CLI as new subcommands.

Reviewer: ```suggestion
- **Legacy CLI plugins**: these are the traditional subprocess plugins that Helm 3 would call as user-specified binaries, and which would extend the Helm CLI as new subcommands.
```

> - **Helm 4 default plugins**: these are of 2 kinds: Chart API v2 and Chart API v3 the default plugins. These should be bundled with Helm 4 so that chart API v2 charts still function exactly like they did in Helm 3.

Author: ok thanks—would explain here but the HIP should be self-explanatory. Will clarify

> - **Helm 4 default plugins**: these are of 2 kinds: Chart API v2 and Chart API v3 the default plugins. These should be bundled with Helm 4 so that chart API v2 charts still function exactly like they did in Helm 3.

Author: @sabre1041 does this make more sense? bbdca4e

> - **Helm 4 default plugins**: these are of 2 kinds: Chart API v2 and Chart API v3 the default plugins. These should be bundled with Helm 4 so that chart API v2 charts still function exactly like they did in Helm 3.

Author: Reworked all of this to be more clear

> - **Helm 4 default plugins**: these are of 2 kinds: Chart API v2 and Chart API v3 the default plugins. These should be bundled with Helm 4 so that chart API v2 charts still function exactly like they did in Helm 3.

Reviewer: Looks good

## High level description of proposed architecture

### Vision
- move to a smaller Helm core, offloading functionality to default plugins, which can be either swapped out or extended with additional plugins to support advanced use cases

### Goals

> - define different types of plugins for each part of existing Helm functionality (we could summarize all in-scope Helm functionality as fitting into different phases of the Helm packaging and release lifecycle)

Author: Absolutely. This is what's meant by this step under "Implementation plan":
> **Define Plugin Interface and Types**: Develop standardized interface for plugin execution, and define the initial plugin types.

I will clarify this. Either by appending to this line: 
```diff
- **Define Plugin Interface and Types**: Develop standardized interface for plugin execution, and define the initial plugin types.
+ **Define Plugin Interface and Types**: Develop standardized interface for plugin execution, and define the initial plugin types (categories of plugins supported by Helm).
```

Or by adding "Plugin types" to the glossary and clarifying there. Whichever feels more clear as I do other cleanup on this HIP wording.

> - define different types of plugins for each part of existing Helm functionality (we could summarize all in-scope Helm functionality as fitting into different phases of the Helm packaging and release lifecycle)

Author: Actually I think this is clear in the first goal at the top:
> - define different types of plugins for each part of existing Helm functionality (we could summarize all in-scope Helm functionality as fitting into different phases of the Helm packaging and release lifecycle)

This is repeated in different ways a few times throughout, so I'll try to consolidate those.

> - define different types of plugins for each part of existing Helm functionality (we could summarize all in-scope Helm functionality as fitting into different phases of the Helm packaging and release lifecycle)

Author: I reworked all of this to be more clear

> - define different types of plugins for each part of existing Helm functionality (we could summarize all in-scope Helm functionality as fitting into different phases of the Helm packaging and release lifecycle)

Reviewer: looks good

- ultimately move as much existing functionality/code as possible out of Helm core into default plugins

>     - this will allow us to more clearly define the core value proposition of Helm (which users must accept otherwise don't use it)

Author: agree, unnecessary

>     - and will let users modify nearly all Helm functionality except that core value proposition

Author: These indented bullets below the goal to "ultimately move as much existing functionality/code as possible out of Helm core into default plugins" are meant to be a list of benefits of doing that. I will clarify this.

>     - and will let users modify nearly all Helm functionality except that core value proposition

Author: I reworked all of this to be more clear

>     - will make helm core more maintainable, and allow plugins to have their own release lifecycle independent of Helm core (eg., http repo downloader plugin can freeze all new features after X date, and supports bug and security fixes for 1 year after Helm is released, sunset on Y date, etc)

Author: Same as https://github.com/helm/community/pull/388#discussion_r1962214056

>     - will support an ecosystem of Helm extension plugin/extension developers, and allow the community of contributors to continue to grow to support end user needs without requiring expanding Helm core

Author: Same as https://github.com/helm/community/pull/388#discussion_r1962214056

> - improve the security model of existing Helm 3 plugin architeecture (now that Wasm is available for sandboxing, and to clearly define scope a plugin of each type can act upon, etc. And now that Extism and XTP are available to make plugin creation, object modeling, and language binding generation more acheivable)

Author: Yes, under "Goals", I agree. Will move this to the "Security implications" section.

> - improve the security model of existing Helm 3 plugin architeecture (now that Wasm is available for sandboxing, and to clearly define scope a plugin of each type can act upon, etc. And now that Extism and XTP are available to make plugin creation, object modeling, and language binding generation more acheivable)

Author: Fixed in bae7382

### Requirements

> - preserve Helm 3 functional backwards-compatibility for Chart API v2 using default plugins, while allowing advanced users to alter that functionality

Author: I meant allowing advanced users to still alter default charts API v2 default functionality via alternative plugins. I will clarify.

> - preserve Helm 3 functional backwards-compatibility for Chart API v2 using default plugins, while allowing advanced users to alter that functionality

Author: Fixed in 213a533

To support this transition, Helm 4 must address the challenges of plugin distribution, ensuring the user installation experience remains as seamless as in Helm 3. It's also important to ensure the proposed architecure works for SDK as well as CLI users.

## Motivation

<!--Clearly explain why the existing design is inadequate to address the problem that the HIP solves.-->

### Why Helm 3 plugins were inadequate to address the problem this HIP solves

<!-- The existing design of Helm 3 plugins is inadequate to address the problem that the HIP solves. -->

This high-level goal of Helm 3 plugins will remain the same for Helm 4 plugins:
> Helm plugins are add-on tools that integrate seamlessly with Helm. They provide a way to extend the core feature set of Helm, but without requiring every new feature to be written in Go and added to the core tool (and maintained by core Helm maintainers)

However, the Helm 3 plugin system has the following limitations:

- Because plugins are primarily intended to add new Helm CLI subcommands, the Helm 3 plugin architecure is not helpful for SDK users or for customizing the existing Helm core functionality.

> - Helm 3 plugins shells out to a user-specified arbitrary binary, which does not allow Helm to limit the scope of these commands.

Author: Added in e8569ae

Note that the primary value proposistion of allowing users to run these binaries as Helm subcommands rather than independently is that Helm-related context is passed, which saves the subcommand binary additional work of fetching that information if it is needed.
- Note that one specialized type of Helm 3 plugin—**downloader plugins**—comes closest to what Helm 4 is modeling its new plugin types after. Downloader plugins are called by Helm for a specific purpose and have a well-defined scope.

### Why other customization options were inadequate to address the problem this HIP solves

Apart from plugins, Helm 3 attempts to allow user customization both globally, and per chart:

- Globally: By overriding Helm environment variables that allow certain configurationl, such as setting the backend storage driver (`HELM_DRIVER`) or the Bearer KubeToken used for authentication (`HELM_KUBETOKEN`). These can be found by running `helm --help` or in the [`helm` command documentation](https://helm.sh/docs/helm/helm/#synopsis).

> - Per command: By setting CLI flags. This is the primary way Helm 3 allows user-defined configuration. It's important to note that these have proliferated over time in order to support backwards-compatibility within each previous MAJOR version of Helm. In addition to the high volume of open issues ane PRs, these flags are clear evidence of the massive need for user-defined customization of Helm. Also by setting the equivalent SDK function options (all SDK functions must support the same configuration as flags since the SDK powers the CLI)

Author: This sentence is awkwardly placed. I meant "Per command" via CLI flags and the equivalent SDK function options. There is just a long note in-between so it doesn't read as it should. Will fix.

> - Per command: By setting CLI flags. This is the primary way Helm 3 allows user-defined configuration. It's important to note that these have proliferated over time in order to support backwards-compatibility within each previous MAJOR version of Helm. In addition to the high volume of open issues ane PRs, these flags are clear evidence of the massive need for user-defined customization of Helm. Also by setting the equivalent SDK function options (all SDK functions must support the same configuration as flags since the SDK powers the CLI)

Author: Fixed in 9eb9e8b

Helm 4 aims to address these limitations by implementing a robust plugin-based system that allows deeper lifecycle integration and modular extension.

To support this transition, Helm 4 must address the challenges of plugin distribution, ensuring the user installation experience remains as seamless as in Helm 3.

## Rationale

*Describe why particular design decisions were made.*

WIP

## Specification

*Describe the syntax and semantics of any new feature.*

### Chart-requied and user-specified plugins

Some plugin types would be optional, such as validation plugins or backend storage plugins, which are up to the discretion of the chart operator's use cases.

However a subset of these plugin types would be required by a chart. Some examples are:
- render plugins (if a chart is written in YAMLScript of Pkl, these plugins are required to actually use the chart)
- downloader plugins—perhaps—for dependencies. If a dependent chart location is specified at, eg., s3:// or git:// then the user would need this plugin to install the chart.

> Chart-required plugins may need to wait for Chart API v3. hile Helm 3 currently **does** allow additional keys to be added ot the Chart struct while retaining Go backwards-compatibility, any keys that add required functionality for a chart to install would not work with older versions of Helm (such as render plugins) and therefore a top level `plugins` key should wait for Chart API v3.

Author: I agree. Will move this from "may" to "will" need to wait for v3 charts 👍

> Chart-required plugins may need to wait for Chart API v3. hile Helm 3 currently **does** allow additional keys to be added ot the Chart struct while retaining Go backwards-compatibility, any keys that add required functionality for a chart to install would not work with older versions of Helm (such as render plugins) and therefore a top level `plugins` key should wait for Chart API v3.

Author: fixed in 588d667da846ee124f191412caefc6c28ebe5919

## Backwards compatibility

<!-- Describe potential impact and severity on pre-existing code. -->

To ease the transition from Helm 3 to Helm 4:
- Existing Helm functionality will continue to work with minimal changes.
- New plugin-based extensions will be optional but encouraged for advanced use cases.
- Helm 4 will include migration guides and tools to assist users.

While the Helm commands and SDK may differ, Chart API v2 charts MUST continue to work as-is (see [Requirements](#requirements) above). Therefore, with this new plugin-first architecture, default plugins (required for Chart API v2 charts to work as they did with Helm 3) must be available to Helm 4 users without needing to do additional work.

For discussion:
- The standard distribution of Helm come bundled with the default plugins for chart api v2 and v3
- advanced users may optionally install Helm without the default plugins if they know they don't need them, for a more streamlined experience
- For SDK users:

>     - For SDK users who do not want to allow filesystem access, potentially Go's embed file functioinality could be used for the default Wasm plugins

Author: Thanks @banjoh for the note about in-memory `embed.FS` and build tags conversation.

```suggestion
    - For SDK users who do not want to allow filesystem access, potentially Go's embed file functionality could be used for the default Wasm plugins. This is a valuable option to explore, especially as `embed.FS` would allow the code to tread either embedded plugins or plugins on disk in the same way. See https://pkg.go.dev/embed. This could also pair nicely with build tags for exclude/include a default plugins directory (eg., default vs contrib directory, etc).
```

    - If filesystem access is an option, but security is still a concern, the SDK user project could allow a specific filesystem location for Wasm plugins. The plugins could then be—at the discretion of the SKD users—either built into a container image, or user-mounted.

## Security implications

<!-- How could a malicious user take advantage of this new feature? -->

- Plugins will run in a sandboxed environment to mitigate security risks.
- Helm will provide guidelines for securely developing and distributing plugins.

Vetting plugins timeline:
1. initially, Helm will only work with plugins developed by and/or under the control of the Helm org until we get the interfaces more fully established.
1. A verification process will be developed to ensure trusted plugins.
1. A process will be established to ensure secure discovery and distribution of plugins.

> 1. we can allow community plugins

Author: I agree. I've removed the verification process line in fb797e5. That was a (very conservative) stopgap for discussion, from a concern that a bad actor chart author may try to take advantage of users testing this out during development of the feature and before a secure process is established for discovery and distribution of plugins. I think that was overkill for this HIP, and yes not realistic for our timeline.

> 1. we can allow community plugins

Reviewer 2: http://ed25519.cr.yp.to/ might work well for this.
https://pkg.go.dev/crypto/ed25519

As I understand it, this allows multiple entities to sign, and consumers can choose their signature requirements.

> note: I've been looking into this as a method for simple YAML file signing.

TBD: Host functions

- Ideas on the above for **discussion**:
    - For community plugins (not under the Helm GitHub org), perhaps the Helm security team could vet and approve specific versions of a plugin (identified by the commit hash) to start. This is not ultimatly sccalable as a job for the Helm team
    - Require signing. Also require verifying of non-helm org (community) plugins. Then the Helm team would need to vet and accept community author signing keys for trusted plugin authors? Or perhaps skip this and go right to the next point:
    - Eventually, plugin consumers (end users) would need to explicitly accept the additional signing keys by non-helm authors.

>     - Artifact Hub's verified and official labels could be used to communicate the status of default plugins. Perhaps something like a "vetted" label could be added?

Author: Matt noted this and I agree that the idea even above these discussion notes is out of scope (I replied in https://github.com/helm/community/pull/388#discussion_r1962544514). For now I removed the label idea a29ebb6 and will soon replace all these open discussion notes with immediately actionable items.

1. **Build Plugin Framework**: Implement the core framework for plugin management and invocation.
1. **Define Plugin Interface and Types**: Develop standardized interface for plugin execution, and define the initial plugin types.
1. **Build Initial default Plugins:** Move Helm 3's Chart API v2 functionality into initial plugins, which will be the default for Chart API v2 charts. 
1. **Ensure Secure Distribution**: Establish a secure process for plugin discovery and distribution.
1. **Develop Documentation**: Publish comprehensive guides for users and plugin developers.

### Good candidates for initial plugins types to implement

- **Chart downloader plugins:**
ie, oci, s3, http-repo as default plugins (Helm 3 functionality). and others optionally such as s3, git, etc.
multiple could be used, for example in dependency charts,
ie, parent chart from oci, child chart from http repo.
- **Validation plugins:**
For example, JSON Schema, CUE validation, CEL, kubeconform, etc.
- **Renderer plugins:**
Maybe more than one in a single chart? based on mime type, eg .yaml .pkl, etc

> ie, gotemplate, CUE, pkl, etc.

Author: fixed

## How to teach this

<!-- How to teach users, new and experienced, how to apply the HIP to their work. -->

WIP

- Documentation that very clear to follow
- Blog posts and conference talks explaining how to take advantage of the new plugin system
- Example plugin templates

## Reference implementation

<!-- Link to any existing implementation and details about its state, e.g. proof-of-concept. -->

We will be using Wasm for Plugin Interfacing, leveraging the Extism project. See <https://extism.org/> and specifically the Extism [Go SDK](https://github.com/extism/go-sdk) and [Go PDK](https://github.com/extism/go-pdk).

We can also use the new XTP for bindings / codegen / host functions https://github.com/extism/go-pdk?tab=readme-ov-file#generating-bindings
    - Note we would still have responsibility to—through the host function—get the pointer. Then Extism makes it easy to readbytes, allocate memory etc. Ultimately, it makes it easy for a plugin implementer to write, but some work in the host function. And the folks at Dylibso are very happy to help us at that stage.
    
### Proof-of-concept
    

> Two example Wasm plugins leveraging Extism have been prototyped. The first is an example downloader plugin, and the second is an example render plugin:

Author: See https://github.com/helm/community/pull/388#discussion_r1962462626

- https://github.com/gjenkins8/helm-plugin-ocigetter
- https://github.com/gjenkins8/helm-plugin-gotemplate-renderer/

The host functions are prototyped using Go testing, in `/testdriver/main_test.go` in each Git repo. These are fairly basic to start, but proves out that this works. Creating the proper interfaces will be part of the work implementing this HIP (see "[Implementation Plan](#implementation-plan)").

### Technical reasoning for Wasm plugins

For a plugin system to work, Helm must be able to “invoke” the plugin with a suitable message so that the plugin can perform its operation and return a result. The traditional "[foreign function interface](https://en.wikipedia.org/wiki/Foreign_function_interface)” problem

> (source: https://kerkour.com/webassembly-wasi-preview2)

Author: This is the image this source note refers to, which helps to explain Wasm plugin invocation (exported and imported functions). Adding as a comment now for discussion. If this feels useful, we can make a version of this diagram in mermaid to include in the HIP.

<img width="730" alt="Screenshot 2025-02-18 at 12 51 37 PM" src="https://github.com/user-attachments/assets/fb46af9b-4649-45e9-8849-6dfb198e4c7a" />

Author: Made a better version in 76a0ea48257847c49b8ada963f49a0acfe5b6615

<img width="580" alt="Screenshot 2025-02-19 at 5 24 29 PM" src="https://github.com/user-attachments/assets/ae59f3f2-d35a-4e91-a7cc-24772c990976" />

> We will be using Wasm for Plugin Interfacing

Author: See https://github.com/helm/community/pull/388#discussion_r1962462626. Will fix as "Wasm" across the board

> We will be using Wasm for Plugin Interfacing
> For a plugin system to work, Helm must be able to “invoke” the plugin with a suitable message so that the plugin can perform its operation and return a result. The traditional "[foreign function interface](https://en.wikipedia.org/wiki/Foreign_function_interface)” problem
> (source: https://kerkour.com/webassembly-wasi-preview2)
> 

Author: removing duplicate paragraph
```suggestion
```

For this, we will be using two of the 3 main technologies exist for interfacing with WASM modules/components today (note the 3rd option—WIT/WASI Preview 2 (WASM IDL)—is under rejeccted ideas at least until it reached further maturity):

- Native Wasip1 (POSIX/C-life FFI)

> Waspi1 defines a way for a module to export basic types in a C/POSIX style FFI. While simple, the caller has to manage much of the details like converting internal structures into simple types (integers, pointers), and (manually) managing memory.

Author: added in 404951a216dac05fb0bd344c4002bd4c8f0c9ff9

- Message serialization/RPC (GRPC/protobuf, JSON, etc)
Rather than trying to implement a rich FFI, either directly via Wasip1, or indirectly via WIT. Implement/utilize a simple FFI that uses a higher level serialization ie. serialization of internal types into e.g. protobufs or JSON and utilizing e.g. a simple FFI or GRPC or HTTP for the majority of message content.

## Rejected ideas

*Why certain ideas that were brought while discussing this HIP were not ultimately pursued.*

WIP

### 🚫 A central repository for plugins

> - A centralized repository for community plugins. Helm moved from a central charts repository to a distributed one [for documented reasons](https://github.com/helm/community/blob/main/hips/archives/helm/distributed-search.md) that we can anticipate applying here—including allowing scaling as the number of charts grew, and ecosystem growth through community-managed alternatives.

Author: Will fix

> - A centralized repository for community plugins. Helm moved from a central charts repository to a distributed one [for documented reasons](https://github.com/helm/community/blob/main/hips/archives/helm/distributed-search.md) that we can anticipate applying here—including allowing scaling as the number of charts grew, and ecosystem growth through community-managed alternatives.

Author: Fixed

- Go native plugin system https://pkg.go.dev/plugin. It does not support Windows

### 🚫 Non-WASM plugins

1. OS/Arch distribtuion problems
2. Security concerns of non-sandboxed execution

These include:
- Go native plugin system
https://pkg.go.dev/plugin
    - downsides:
        - does not support Windows. For this reason alone, we can't use it
        - also the version of go, architecture, and (some other build options) MUST match between parent and plugin
        - this means WHEN helm core go version upgrades, none of the community plugins will work until the plugin maintainers upgrade their plugin to match the new helm core Go version
- Hashcorp's go-plugin package
https://github.com/hashicorp/go-plugin
    - this is also what protobuf does (subprocess)

> - embedded go interpreter

Reviewer: ```suggestion
- Embedded go interpreter
```

    - traefik does this (github.com/traefik/yaegi)
- Caddy novel Go plugin approach
https://caddyserver.com/docs/architecture

### 🚫 Non-Extism WASM plugins

- wazero (directly)
https://github.com/tetratelabs/wazero
    - note that [extism/go-sdk](https://github.com/extism/go-sdk) (see Extism under [Reference implementation](#reference-implementation)) leverages wazero in a viable way.
    - downsides:

>         - all params are uint32. So any strings must be interpreted from that

Reviewer: ```suggestion
        - all params are uint32. So, any strings must be interpreted from that
```

> or in our case structs must be converted in/out (may need to write a library for this?)

Reviewer: ```suggestion
or in our case, structs, must be converted in/out (may need to write a library for this?)
```

>         - The allocation example uses CGO to allocate memory, uses unsafe package. note we don't need to use tinygo

Reviewer: Review this sentence as it sounds awkward (use of uses)

Author: It does sound weird, it refers to the https://pkg.go.dev/unsafe package, not just the word "unsafe", will clarify

Author: Actually will just remove this entire line since it was from initial findings and is just one of their examples, not necessary for using wazero.

- Wadge in wasmCloud
https://wasmcloud.com/docs/developer/languages/go/components#wadge
    - Bailey Cosmonic CTO suggested (via Taylor Thomas)
    - downsides:
        - relies on CGO
- Envoy's Wasm approach
https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/wasm_filter
- WIT/WASI Preview 2 (WASM IDL)

>     - WASI P2 defines WIT as a IDL for wasm modules components. Greatly simplifying message parsing capabilities. However, as WASIP2 is (very) new, support is quite limited (unsupported in wazero for example).

Reviewer: ```suggestion
    - WASI P2 defines WIT as a IDL for WASM modules components which greatly simplifying message parsing capabilities. However, as WASIP2 is (very) new, support is quite limited (unsupported in wazero for example).
```

Author: needed an additional change. Fixed in 332a9826048208293dfe9dfb96a5556539e18a97

## Open issues

*Any points that are still being decided/discussed.*

TBD

## Future Work

- Expanding plugin functionality to support additional lifecycle phases in addition to the "good candidates for initial plugins" above.
- Developing advanced tools for debugging and testing plugins.

> - Investigate the need to create migration tools to help users transition from Helm 3 to the Helm 4 plugin-first architecure (or from Chart API v2 to v3). For prior work see the [helm-2to3 plugin](https://github.com/helm/helm-2to3).

Reviewer: ```suggestion
- Investigate the need to create migration tools to help users transition from Helm 3 to the Helm 4 plugin-first architecture (or from Chart API v2 to v3). For prior work see the [helm-2to3 plugin](https://github.com/helm/helm-2to3).
```

## Acknowledgments

WIP

- Special thanks to the Helm community for their feedback and contributions to this proposal.
 
<!-- We can already say:-->
- Thanks to Extism and Wazero maintainers for guidance

<!-- Ideally we'll also be able to say:

> - Thanks also to Fermyon, Cosmonic, and Dylibso for guidance and collaboration on Wasm integration in Helm.

Reviewer: ```suggestion
- Thanks also to Fermyon, Cosmonic, and Dylibso for guidance and collaboration on WASM integration in Helm.
```

Author: Should be Wasm not WASM, even though the doc already contains lots of the latter. Will fix all of these after I commit your other applicable suggestions.

Reviewer: Yup. Wasm is the correct spelling

---
hip: 9999
title: "H4HIP: Wasm plugin system"
authors: [ "Scott Rigby <scott@r6by.com>", "George Jenkins <gvjenkins@gmail.com>" ]
created: "2025-02-17"
type: "feature"
status: "draft"
helm-version: "4"
---

## Abstract

Helm is currently a monolithic application that is difficult to customize without changing the core codebase. This requires maintainers to review and accept every contribution, which is time-consuming and not scalable.

This architectural proposal outlines a concrete plan to define a wider range of plugin types in Helm 4, so that users will be able to customize more default functionality in their own environment through alternative plugins. This will be accomplished through a new Wasm based plugin system.

This will result in making Helm not only more extensible, but also more maintainable, allowing contributors to extend more of Helm's functionality via plugins rather than core changes to meet their needs more quickly.

## Scope

> This HIP will cover the overall vision for the new plugin system, and an [implementation plan](#Implementation-Plan) for initial plugin types introduced into this system. This will will allow more categories of Helm functionality to become extended through additional plugin types in the future.

Reviewer 3: It might be me, but how are we handling existing plugins?

- Per command: By setting CLI flags and the equivalent SDK function options. This is the primary way Helm 3 allows user-defined configuration. It's important to note that command flags have proliferated over time in order to support backwards-compatibility within each previous MAJOR version of Helm. In addition to the high volume of open issues and PRs, these flags are clear evidence of the massive need for user-defined customization of Helm.
- Per chart: By defining Hook annotations in a chart's Kubernetes resources, Chart authors can trigger Kubernetes objects at specific lifecycle phases (e.g., pre-install, post-upgrade)

While useful, all of these configuration approaches are limited and do not provide the level of extensibility users have requested without requiring their use cases to be built into Helm core.

## Specification

*Describe the syntax and semantics of any new feature.*

> We will implement a Wasm plugin architecture in Helm 4 that componentizes core Helm functionality into discrete, replaceable plugins. The ultimate goal is to move as much existing functionality/code as possible out of Helm core into default plugins.

Reviewer: Recommend avoiding 1st person point of view

Author: Do you have another suggestion than using "we" from the point of view of the Helm project? A number of other HIPs use this same phrasing too—I'm honestly not sure of a clearer way to say it. Do you have suggestions of a different way to say this throughout the HIP?

Reviewer: I always tend to favor third person. Something like "A Wasm plugin architecture will be implemented in Helm 4..."

Reviewer: Yup. Looks great

### Four kinds of plugins

There will be four kinds of plugins in the new system: Default Behavior plugins, Alternative Behavior plugins, Chart-defined plugins, and CLI plugins.

#### 1. Default behavior plugins

These plugins will be used to implement the default behavior of Helm that can then be overridden by Alternative Behavior plugins. This includes all functionality for Chart API v2, non-chart-specific functionality for Chart API v3, and other Helm functionality not tied to a Chart ApiVersion. Because of this, these plugins will be automatically loaded into the default installation of Helm 4.

> Like all other Helm 4 plugins, these will be written as Wasm plugins. However, while non-default plugins may be written in any language supported by Extism PDK, Default Behavior plugins will be written in Go (effective copied from existing Helm code), in order to keep Helm source codebase—core and default plugins—consistent and more maintainable than a polyglot codebase.

Reviewer: Provide link/reference to Extism PDK

Author: Will do

Author: done

Note that Helm 3 allowed a similar forwards-compatibility for Chart API v2, for example when support for a new gotemplate/sprig function was added to a specific Helm version. Charts could check the Helm version in their template or a helper function, and exit with a friendly error if the version of Helm did not meet this version constraint. We may be able to do better as this applies to plugins in Helm 4.

A first pass for Chart-defined plugin syntax in `Chart.yaml`:

```yaml
# made up example URLs—these are not actual Helm 4 plugins yet
plugins:
    render:
        - https://github.com/pkl-community/helm-pkl@0.1.0: [".pkl", ".pickle"]
        - https://github.com/kubeys/helmys@0.1.4: [".ys"]
    download:
        - https://artifacthub.io/packages/helm-plugin/s3/s3@0.16.3: ["s3"]
        - https://artifacthub.io/packages/helm-plugin/git/helm-git: ["git", "git+https", "git+ssh", "git+file"]
```

#### 4. CLI plugins

> The new plugin system will continue to allow users to build and install CLI plugins, which specify subcommands for the Helm CLI. However, these will also be run in Wasm sandbox rather than directly calling arbitrary executables as a subprocess as we done in Helm 3. This will require Helm 4 plugin developers to follow our tutorials for migrating their legacy CLI plugin to the new Wasm CLI plugin framework.

Reviewer: ```suggestion
The new plugin system will continue to allow users to build and install CLI plugins, which specify subcommands for the Helm CLI. However, these will also be run in Wasm sandbox rather than directly calling arbitrary executables as a subprocess as was implemented in Helm 3. This will require Helm 4 plugin developers to follow documentation which will be created detailing the process for migrating legacy CLI plugins to the new Wasm CLI plugin framework.
```

---
hip: 9999
title: "H4HIP: Improve and extend plugin system"
authors: [ "Scott Rigby <scott@r6by.com>", "George Jenkins <gvjenkins@gmail.com>" ]
created: "2025-02-17"
type: "feature"
status: "draft"
helm-version: "4"
---

## Abstract

**A short (~200 word) description of the technical issue being addressed.**

Helm is currently a monolithic application that is difficult to customize without changing the core codebase. This requires maintainers to review and accept every contribution, which is time-consuming and not scalable.

This architectural proposal outlines a concrete plan to build a new Wasm plugin system and define a wider range of plugin types in Helm 4 to componentize as much default functionality as possible. Users will be able to customize the default functionality in their own environment through alternative plugins.

> This will result in making Helm more maintainable and extensible, improve security through Wasm sandboxing, and allow contributors to extend Helm's functionality via plugins rather than core changes to meet their needs more quickly.

Reviewer 4: I don't believe this is going to make Helm more maintainable. Here are my initial reasons:

1. Instead of one Helm binary to build per os/arch we are going to have multiple components to build and distribute. This increases distribution complexity.
2. We will have to document and answer questions on how to build Helm in the new setup. The Helm project is not the only builder/distributor of Helm. Linux distributions and Homebrew are two more examples. Looking through docs on Homebrew, I don't see anything on building wasm. I expect difficulties onboarding other distributors to the new setup.
3. Increased build/release times. This will impact the person doing the releases. I can understand the theory of components being released at different times but the reality is that we rely on too few people to ensure releases go out so I expect the person doing the releases will need to ensure all the components are property released and in the build.
4. Instead of being direct Go calling Go there are going to be more complex interactions. This is doing to increase test complexity for the whole combined system.

Reviewer 5: As a Flux maintainer this proposal terrifies me, I can't imagine how are we going to make helm-controller work with this new design. We'll have to build the plugins as part of Flux CI pipeline and match their versions to the Helm SDK version, embed the wasm binaries in the Flux containers and somehow prevent random charts downloading and executed untrusted code which could break the multi-tenancy enforcement we have in place. I feel like this design will break all controllers currently using the Helm SDK.

Author: We have updated this HIP so that plugins are completely optional, both for CLI users, and no need for SDK users to deal with it if they don't want to.

Author: Also, the chart-defined plugins idea has moved to a separate, follow-up HIP so that these concerns can be addressed separately from this HIP.

    862HB19B    gitlab.com/someone/s3-helm-plugin
    208DD36E    github.com/helm/helm
    208DD36E    github.com/helm/helm
    774UR67F    github.com/someone/cue-helm-plugin
    208DD36E    github.com/helm/helm
    no          local-dev
    ```
- `helm plugin install` will now install the latest release from a VCS repo URL, if a release version is not specified with `--version`
- `helm plugin update` will check for updates, then follow the same behavior of `plugin install`
- `helm plugin sign` will use Helm 4's built-in provenance signing mechanisms as a convenience to sign a plugin
- `helm plugin verify` will use Helm 4's built-in provenance mechanisms as a conveniencce to verify a signed plugin
- `helm plugin create` will add create a boilerplate plugin scaffolding, specified by `--type`. `--help` will list the types of plugins supported by your installed version of Helm

Additionally commands will be added to allow packaging Chart-defined plugins within the chart (see [Distribution](#Distribution) for reasoning):

- `helm chart plugin build` similar to `helm dependency build` but for plugins
- `helm chart plugin update` similar to `helm dependency update` but for plugins

> ### Distribution

Reviewer 4: Why wouldn't we support the downloader plugins as the means to download WASM? That would provide http(s) and OCI out of the box. It would make it extensible in the same way everything else is.

Author: Absolutely want this. It is a goal, if we can implement that in time. Otherwise, we can start with the current options and do this as a follow up.

```mermaid
sequenceDiagram
    participant h as Helm (Go) + Wasm runtime
    participant P as Plugin (Go, Rust, etc as .wasm)
    h->>P: Call exported function
    P->> h: Call imported function
```

For this, Helm 4 will be using 2 of the 3 main technologies that exist for interfacing with Wasm modules/components today (the 3rd option—WASI P2—is under [Rejected ideas](#Rejected-ideas)). The two that will be used here are:

1. Native WASI P1 (POSIX/C-life FFI)
[WASI P1](https://wasi.dev) defines a way for a module to export basic types in a C/POSIX style FFI. While simple, the caller has to manage much of the details, like converting internal structures into simple types (integers, pointers) and (manually) managing memory.
2. Message serialization/RPC (GRPC/protobuf, JSON, etc)
Rather than trying to implement a rich FFI, either directly via WASI P1, or indirectly via WIT. Implement/utilize a simple FFI that uses a higher level serialization ie. serialization of internal types into e.g. protobufs or JSON and utilizing e.g. a simple FFI or GRPC or HTTP for the majority of message content.

### Technical reasoning using Extism for Wasm plugins

> For Wasm Plugin Interfacing, Helm 4 will be leveraging the Extism project. See <https://extism.org/> and specifically the Extism [Go SDK](https://github.com/extism/go-sdk) and [Go PDK](https://github.com/extism/go-pdk).

Reviewer 4: Extism is a project by Dylibso. They are a VC funded startup whose last round was a seed round in 2023. Given how many startups pivot or shut down, it's a high risk to tightly couple to their system as the following paragraphs describe.

I was expecting we would use a runtime behind an interface (think adapter pattern). This way, if there was a problem with the runtime (e.g., it became unmaintained) we could exchange it for another implementation. This reduces risk on Helm.

We ran into this problem with ORAS when it did a v2 design but we had tightly coupled to the v1 design. We were lucky that the ORAS project has been willing to work with us. I don't expect we'll get this with WASM runtimes from VC startups.

This would mean that the plugin system needs to be something tied to the WASM spec rather than a specific implementation and the rest filled in by Helm.

I bring all of this us due to risk management.

Reviewer 6: Agree on designing with an adapter pattern in mind. I do think that if/when Wazero catches up to other Wasm runtimes (supporting WASI p2 and p3) then it is likely that WIT will be both available and desirable as the bindings language. And I don't see any downsides to using extism right away. (I believe it will also continue to track mainline Wasm over time.)

        The only other viable Component model alternative to Extism for our use case that we're aware of would be to use CGo to link in [Wasmtime](https://github.com/BytecodeAlliance/wasmtime), but that would likely require a high level of expertise to implement and maintain.

    - [Wazero](https://github.com/tetratelabs/wazero) directly
    
        Problem: With Wazero, all params are uint32, so any strings must be interpreted from that, which would be a significant amount of work to manage directly. This is one of the problems Extism solves, and is able to leverage Wazero in a viable way for our plugins use case.

    - [Wadge](https://wasmcloud.com/docs/developer/languages/go/components#wadge) in wasmCloud
    
        Problem: This is another CGo-based solution. Helm maintainers would like to avoid CGo due to complexity it would add for cross-compiling, debugging, and other issues.

>     - [Web Assembly for Proxies (Go SDK)](https://github.com/proxy-wasm/proxy-wasm-go-sdk)

Reviewer 4: cgo introduces a distribution problem. How does Helm distribute the wasm runtime or ensure it's there is a cross-environment manner (i.e. windows, linux, mac with different architecture and distributions of Linux)?

Author: Agree. This line, and anything with CGo dependency is in under the "Rejected ideas" category.

Reviewer 6: In my view, `wasmtime` is the best runtime available. However, it is Rust, so would have to be `cgo`, and I still stand by my view that cgo almost always introduces more problems than it solves. It is too bad that wazero is so far behind the other runtimes in terms of WASI support, but it will catch up over time, at which point the benefits of the Component Model will be realized anyway. So I think we should consider ourselves to be "playing the long game" by choosing a Go-native implementation even if it is technically inferior in the short term.

Reviewer 2: In my opinion, the CGO issue (and not just for Helm, but the entire Go landscape) has a fairly easy solution.
From all my research and conversations (including with @mattfarina, @technosophos (just a few minutes ago) and others), CGO presents 1 main problem and 1 secondary problem.

The main problem is distribution of 2 separate components: the Go compiled binary file and the foreign shared library binary file it binds to.
The secondary problem is integrating the foreign build system into a Go project.

Go has the [embed](https://pkg.go.dev/embed) package that lets you embed any external file into a Go binary, so that you can ship needed assets for a piece of Go software.
That means we can simply embed the wasmtime shared library into the helm binary for every platform build.
The needed shared library build will always match the needed helm build far every targeted platform.

I'm working on a project to make all of this trivial to add to a Go build system. I'll report back on this soon.

The main point is that wasmtime is really the best solution for now, and seems like it will be for some time.

If the only thing holding back wasmtime is CGO, I'm pretty sure we can solve this.

Reviewer 6: I love this approach! It would actually be a great general solution (and that alone excites me), but the idea that we could have Wasmtime as the engine actually has a few really big advantages:

* First and foremost, support for the latest Wasm WASI standards
* Because Rust's memory manager is far more efficient, the actual performance should be better (which would impact Python and JS disproportionately
* Wasmtime is maintained by a foundation, which would meet @mattfarina 's stability criteria.

Eventually, it is recommended that Helm supports plugin distribution via OCI artifacts. And further distribtion schemes e.g. S3 could be supported via plugins themselves.

> To enable airgapped installations and other use cases where downloading Chart-defined plugins is not feasible, plugins may be packaged and distributed with the chart itself, similar to how Helm 3 allows this for a chart's subchart dependencies. A command will be added to enable this (see [Helm plugin commands](#Helm-plugin-commands)).

Reviewer 7: The distribution via OCI and mirroring / aliasing of OCI registries would also go a long way to solving for air-gaped environments.

Bundling plugins with charts has the downside that charts will become much larger (go template render is 19MiB uncompressed / ~3MiB compressed for example). But otherwise bundling is completely possible of course.
